src/librustc_middle/ty/layout.rs

```diff
@@ -875,56 +894,124 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     let mut niche_variants = VariantIdx::MAX..=VariantIdx::new(0);
+                    let mut max_size = Size::ZERO;
+                    let mut second_max_size = Size::ZERO;
+                    let mut align = dl.aggregate_align;
+
+                    // The size computations below assume that the padding is minimum.
+                    // This is the case when fields are re-ordered.
+                    let struct_reordering_opt = !def.repr.inhibit_struct_field_reordering_opt();
+                    if struct_reordering_opt {
+                        let mut extend_niche_range = |d| {
+                            niche_variants =
+                                *niche_variants.start().min(&d)..=*niche_variants.end().max(&d);
+                        };
 
-                    // Find one non-ZST variant.
-                    'variants: for (v, fields) in variants.iter_enumerated() {
-                        if absent(fields) {
-                            continue 'variants;
+                        // Find the largest and second largest variant.
+                        for (v, fields) in variants.iter_enumerated() {
+                            if absent(fields) {
+                                continue;
+                            }
+                            let mut size = Size::ZERO;
+                            for &f in fields {
+                                align = align.max(f.align);
+                                size += f.size;
+                            }
+                            if size > max_size {
+                                second_max_size = max_size;
+                                max_size = size;
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = Some(v);
+                            } else if size == max_size {
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = None;
+                                extend_niche_range(v);
+                            } else {
+                                second_max_size = second_max_size.max(size);
+                                extend_niche_range(v);
+                            }
                         }
-                        for f in fields {
-                            if !f.is_zst() {
-                                if dataful_variant.is_none() {
-                                    dataful_variant = Some(v);
-                                    continue 'variants;
-                                } else {
-                                    dataful_variant = None;
-                                    break 'variants;
+                    } else {
+                        // Find one non-ZST variant.
```

Comment:
Here I kept the same logic as before for the rare cases where we want the niche optimization, but the field re-ordering optimization is disabled.  The reason i used the same logic is because it is faster as it does not attempt to estimate the size at all since it cannot be done.  But maybe that is not worth it and it would be simpler to reuse the same code and bail out if second_max_size > 0

Reply:
> where we want the niche optimization, but the field re-ordering optimization is disabled

That should never happen, the two should be tied together.

Reply:
Right, the only way this can happen is if IS_LINEAR is set, which can only happen if running out of "optimization fuel", so not something that should happen, but still could happen.
I simplified the code a bit.